ulong_extras/is_perfect_power.c

```diff
@@ -0,0 +1,154 @@
+                               { 0x2 , 0x0 , 0x20010000 , 0x0 , 0x0 ,
+                               0x1001 }, { 0x2 , 0x3000 , 0x0 , 0x0 , 0x0 ,
+                               0x0 , 0x0 , 0xc000 , 0x4000001 },
+                               { 0x2 , 0x0 , 0x0 , 0x401 }, { 0x2 , 0x0 ,
+                               0x8000001 , 0x0 , 0x0 , 0x8000000 , 0x0 ,
+                               0x18 , 0x0 , 0x0 , 0x0 , 0x0 , 0x0 , 0x0 ,
+                               0x0 , 0x6 , 0x400 , 0x0 , 0x0 , 0x420 , 0x0
+                               , 0x0 , 0x11 }, { 0x2 , 0x0 , 0x180000 ,
+                               0x0 , 0x0 , 0x0 , 0x0 , 0x0 , 0x18000000 ,
+                               0x0 , 0x0 , 0x4001 }
+                              };
+#else
+    mp_limb_t qclass[18][15] = { { 0x3 }, { 0x43 }, { 0x403 },
+                                  { 0x10021003 } , { 0x400003 },
+                                  { UWORD(0x10000040800003) },
+                                  { UWORD(0x300c00000000002) ,
+                                  UWORD(0x4000000001) },
+                                  { UWORD(0x2008000000082) ,
+                                  UWORD(0x200000040000) ,
+                                  UWORD(0x4100000001004001) },
+                                  { UWORD(0x400000000003) },
+                                  { UWORD(0x400000000000003) },
+                                  { UWORD(0x10001000000042)
+                                  , UWORD(0x80000000) , 0x0 , 0x1000000 ,
+                                  UWORD(0x42000000080009) },
+                                  { UWORD(0x100000000002) ,
+                                  UWORD(0x20000000000) , 0x100001 },{ 0x2 ,
+                                  0x40001 }, { 0x2 , 0x20010000 ,
+                                  UWORD(0x100000000001) }, {
+                                  UWORD(0x300000000002) , 0x0 ,
+                                  0x0 , UWORD(0xc00000000000) , 0x4000001 },
+                                  { 0x2 , UWORD(0x40000000001) }, { 0x2 ,
+                                  UWORD(0x108000000) ,
+                                  UWORD(0x800000000000000) ,
+                                  UWORD(0x1800000000) , 0x0 , 0x0 , 0x0 ,
+                                  UWORD(0x600000000) , 0x400 ,
+                                  UWORD(0x42000000000) ,
+                                  0x0 , 0x11 },  { 0x2 , 0x180000 , 0x0 ,
+                                  0x0 , 0x18000000 , UWORD(0x400000000001) }
+                                };
+#endif
+
+
+int n_is_perfect_power(mp_limb_t * r, mp_limb_t x)
+{
+    /*
+       p[i], is prime less than 64.
+       q[i], is prime such that q[i] mod p[i] = 1.
+       if x = y ^ p[i] , for some y than,
+       x ^ ((q[i] - 1) / p[i]) mod q[i] <= 1.
+       see this following link for info:
+       ("http://citeseerx.ist.psu.edu/viewdoc/
+         download?doi=10.1.1.108.458&rep=rep1&type=pdf")
```

Comment:
AKA
Eric Bach, Jonathan Sorenson,
Sieve algorithms for perfect power testing,
Algorithmica 9 (1993), 313–328. MR 94d:11103